packages/nlp/test/nlp.test.js

```diff
@@ -788,6 +788,97 @@ describe('NLP', () => {
     });
   });
 
+  describe('Process an utterance with actions', () => {
+    test('The action is executed when intent matches', async () => {
+      const nlp = new Nlp({
+        languages: ['en'],
+        autoSave: false,
+      });
+      nlp.addDocument('en', 'Who am i?', 'who_am_i');
+      let actionCalled = false;
+      nlp.addAction('who_am_i', 'testaction', ['param1'], (data, param1) => {
+        expect(param1).toEqual('param1');
+        actionCalled = true;
+      });
+      await nlp.train();
+      const input = {
+        locale: 'en',
+        utterance: 'Who am i?',
+      };
+      const output = await nlp.process(input);
+      expect(output.utterance).toEqual(input.utterance);
+      expect(output.intent).toEqual('who_am_i');
+      expect(output.answer).toBeUndefined();
+      expect(actionCalled).toEqual(true);
+    });
+    test('The action is executed after answer generation (default) and answer returned in object', async () => {
+      const nlp = new Nlp({
+        languages: ['en'],
+        autoSave: false,
+      });
+      nlp.addDocument('en', 'Who am i?', 'who_am_i');
+      nlp.addAnswer('en', 'who_am_i', 'You are HAL');
+      nlp.addAction('who_am_i', 'testaction', ['param1'], (data, param1) => {
+        expect(param1).toEqual('param1');
+        data.answer = 'answer';
+        return data;
+      });
+      await nlp.train();
+      const input = {
+        locale: 'en',
+        utterance: 'Who am i?',
+      };
+      const output = await nlp.process(input);
+      expect(output.utterance).toEqual(input.utterance);
+      expect(output.intent).toEqual('who_am_i');
+      expect(output.answer).toEqual('answer');
+    });
+    test('The action is executed after answer generation (default) and answer returned as string', async () => {
+      const nlp = new Nlp({
+        languages: ['en'],
+        autoSave: false,
+      });
+      nlp.addDocument('en', 'Who am i?', 'who_am_i');
+      nlp.addAnswer('en', 'who_am_i', 'You are HAL');
+      nlp.addAction('who_am_i', 'testaction', ['param1'], (data, param1) => {
+        expect(param1).toEqual('param1');
+        return 'answer';
+      });
+      await nlp.train();
+      const input = {
+        locale: 'en',
+        utterance: 'Who am i?',
+      };
+      const output = await nlp.process(input);
+      expect(output.utterance).toEqual(input.utterance);
+      expect(output.intent).toEqual('who_am_i');
+      expect(output.answer).toEqual('answer');
+    });
+    test('The action is executed before answer generation (when configured that way) and answer is not overridden', async () => {
```

Comment:
This test is almost equal than the 2nd you added. If you want to test executeActionsBeforeAnswers add cases who clearly explain the difference between setting it to true, false &  validate that using default value is being understood as false value.
Basically, with value true allow answers to be rendered with entities filled from slot filling. With false, show how this fails.

Reply:
Yes you are right ... I will rework that

Reply:
Ahh I see it now ... tests did not followed all feature additions :-) I added the option that a answer set in an action executed before determine answers is used and the answer determination is skipped. I fix the tests

Reply:
tests adjusted and enhanced to show the power of the new feature :-)